src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -36,43 +33,38 @@ internal sealed class FileChangeWatcher
         /// For performance and correctness reasons, NOTHING should ever do a block on this; figure out how to do your work without a block and add any work to
         /// the end of the queue.
         /// </summary>
-        private Task<IVsAsyncFileChangeEx> _taskQueue;
-        private static readonly Func<Task<IVsAsyncFileChangeEx>, object, Task<IVsAsyncFileChangeEx>> _executeActionDelegate =
-            async (precedingTask, state) =>
-            {
-                var action = (Func<IVsAsyncFileChangeEx, Task>)state;
-                await action(precedingTask.Result).ConfigureAwait(false);
-                return precedingTask.Result;
-            };
-
-        public FileChangeWatcher(Task<IVsAsyncFileChangeEx> fileChangeService)
-            => _taskQueue = fileChangeService;
+        private readonly AsyncBatchingWorkQueue<WatcherOperation> _taskQueue;
 
-        private void EnqueueWork(Func<IVsAsyncFileChangeEx, Task> action)
+        public FileChangeWatcher(
+            IAsynchronousOperationListenerProvider listenerProvider,
+            Task<IVsAsyncFileChangeEx> fileChangeService)
         {
-            lock (_taskQueueGate)
-            {
-                _taskQueue = _taskQueue.ContinueWith(
-                    _executeActionDelegate,
-                    action,
-                    CancellationToken.None,
-                    TaskContinuationOptions.None,
-                    TaskScheduler.Default).Unwrap();
-            }
+            _fileChangeService = fileChangeService;
+            _taskQueue = new AsyncBatchingWorkQueue<WatcherOperation>(
+                TimeSpan.Zero,
```

Comment:
curious why zero, and do we still get abtching benefit from that?  would it be problematic to even go to 15ms to really be able to collect a ton of items to process?

Reply:
Based on the long queues we've observed in the past, it seems likely that we would get benefits even with the short delay. A batching delay of zero matches the behavior we have today, where long chains have been occurring. I plan to keep 0-1ms for this PR, and then we can consider changes in a follow-up.

Reply:
That seems ok to me.  it's likely that just with the thread bouncing that we still aggregate at least a few items into a single batch.  would it be possible for you to confirmt hat though with a little mnual test + logging?  if too complex to do, that's ok.

Reply:
There were a few cases where operations ended up not getting batched during periods of low activity, but this is what ProcessBatchAsync looks like during solution close:

```
[27844] Processing a batch of 36 work items.
[27844] Processing a batch of 276 work items.
[27844] Processing a batch of 285 work items.
[27844] Processing a batch of 285 work items.
[27844] Processing a batch of 2 work items.
[27844] Processing a batch of 8 work items.
[27844] Processing a batch of 3228 work items.
[27844] Processing a batch of 3930 work items.
[27844] Processing a batch of 3523 work items.
[27844] Processing a batch of 633 work items.
[27844] Processing a batch of 336 work items.
[27844] Processing a batch of 2 work items.
[27844] Processing a batch of 301 work items.
[27844] Processing a batch of 2 work items.
[27844] Processing a batch of 3 work items.
[27844] Processing a batch of 169 work items.
[27844] Processing a batch of 129 work items.
[27844] Processing a batch of 2 work items.
[27844] Processing a batch of 284 work items.
[27844] Processing a batch of 2 work items.
[27844] Processing a batch of 15 work items.
[27844] Processing a batch of 284 work items.
[27844] Processing a batch of 330 work items.
[27844] Processing a batch of 303 work items.
[27844] Processing a batch of 660 work items.
[27844] Processing a batch of 1611 work items.
[27844] Processing a batch of 3320 work items.
[27844] Processing a batch of 3256 work items.
[27844] Processing a batch of 2192 work items.
[27844] Processing a batch of 4152 work items.
[27844] Processing a batch of 2483 work items.
```

Reply:
wfm.  i wouldnt' be opposed to a comment stating that emperical testing showed .Zero still provides a huge amoutn of batchign in practice.

Reply:
➡️ Added a comment indicating this.